add icebergs to namelist.config and turn off icebergs by default

ackerlar opened the pull request:
I added default iceberg values in namelist.config and fesom-2.0.yaml and fesom-2.1.yaml. Icebergs are turned off by default, however, the "icebergs" section needs to be present in namelist.config. If it is not, fesom-gamma [https://gitlab.dkrz.de/FESOM/fesom2/-/commits/paleodyn/awiesm-2.1-gamma](url) will crash.

ackerlar committed:
add icebergs to namelist.config and turn off icebergs by default in fesom yaml files

mandresm commented:
Hi @ackerlar ,

I will review this on Monday, then I will suggest some minor changes that need to be done in order to keep the FESOM files a bit more clean.

But I'd like to ask you now about FESOM versioning and include @pgierz into the conversation:
Are the icebergs version of FESOM ever going to be merge into the main branches of 2.0 and 2.1? If not, then we can only add the namelist changes for icebergs behind a choose block for the versions of fesom that contain that option, otherwise these changes will break all the other versions of FESOM that do not support icebergs.

Once I know the answer to that question I'll look into the fesom files and do a PR to `fix/icb_fesom_namelist` with my suggestions.

ackerlar commented:
@mandresm To my understanding, they shall go into the main branch. They are already merge into [[https://gitlab.dkrz.de/FESOM/fesom2/-/commits/paleodyn/awiesm-2.1-gamma](https://github.com/esm-tools/esm_tools/pull/url)], which is, again from my understanding, the FESOM version we are aiming to merge everything into. Icebergs, wiso and pico are already in there, the only thing missing is recom.

mandresm commented:
> @mandresm To my understanding, they shall go into the main branch. They are already merge into [[https://gitlab.dkrz.de/FESOM/fesom2/-/commits/paleodyn/awiesm-2.1-gamma](https://github.com/esm-tools/esm_tools/pull/url)], which is, again from my understanding, the FESOM version we are aiming to merge everything into. Icebergs, wiso and pico are already in there, the only thing missing is recom.

I see, but again fesom-2.1-gamma is not the main branch for everyone. For example, that is not the version AWICM3 uses, and is not the version for standalone FESOM-2.1. Therefore, I think we need to put the namelist_changes for icebergs behind a `choose_version: -> fesom-2.1-gamma`, or add default namelists inside `namelist/fesom/2.1/gamma` and point at this as `namelist_dir` when in version `fesom-2.1-gamma`.

pgierz commented:
Only briefly, I can look in detail next week: In my view, this should reside behind a top level switch, just like wiso and recom.

pgierz commented:
And (sorry for the spam) this fesom awiesm gamma simply needs to be renamed. The fesom version Lars mentioned is the one everyone should switch to (with at some point recom added on top)

mandresm commented on configs/components/fesom/fesom-2.0.yaml:
```diff
@@ -203,14 +216,40 @@ outdata_targets:
-            ib_num_old: 0
-            ib_num_new: 0
-
-            add_restart_in_files:
-                    icb_restart: icb_restart
-
-            add_restart_out_files:
-                    icb_restart: icb_restart
```
Was there a reason why this was moved out from here and placed in the `restart_in_files` out of the `choose_use_icebergs`? Like it did not work otherwise?

mandresm commented on configs/components/fesom/fesom-2.0.yaml:
```diff
@@ -203,14 +216,40 @@ outdata_targets:
-
-            add_restart_out_files:
-                    icb_restart: icb_restart
+            outputs: 
+                    [Av, Bo, Kv, MLD1, MLD2, N2, Redi_K, a_ice, alpha, atmice_x, atmice_y, atmoce_x, atmoce_y, beta, bolus_u, bolus_v, bolus_w, evap, fer_C, fer_K, fh, fw, hbl, iceoce_x, iceoce_y, m_ice, m_snow, prec, reso, runoff, salt, slope_x, slope_y, slope_z, snow, ssh, sss, sst, temp, tx_sur, ty_sur, u, uice, v, vice, vve, w, h2o16, h2o18, hDo16, h2o16_ice, h2o18_ice, hDo16_ice, ibfwe, ibfw ...
```
The `outputs` variable does not seem to be useful anymore (it is not used anywhere in the `fesom-2.0.yaml`). See this:
```
212 outdata_sources:
213         fesom: "*.fesom.*.nc"
214 outdata_targets:
215         fesom: "*.fesom.*.nc"
```
@ackerlar , are you okay with me removing the `outputs` variable, or am I missing something?

mandresm commented:
@ackerlar , are icebergs only supported in `FESOM-2.1`? If that's the case we can get rid of all the icebergs configurations in `fesom-2.0.yaml` to avoid confusion. What do you think?

ackerlar commented:
@mandresm Thank you for your comments! I addressed all your remarks and currently try to make a test run on mistral. But it seems to take a while until it will start...

With respect to the FESOM-2.0/FESOM-2.1 comment: I guess, everyone interested in using icebergs would use FESOM-2.1. I can remove all the iceberg related code from the FESOM-2.0 config yaml. I just used it because esm_tools used to refer to the fesom-2.0.yaml even when fesom-2.1 was used. But I think you have tackled this already, right?

mandresm commented:
> I just used it because esm_tools used to refer to the fesom-2.0.yaml even when fesom-2.1 was used. But I think you have tackled this already, right?

That is correct, `fesom-2.1.yaml` is used now used for all the `FESOM-2.1` variants.

pgierz reviewed (commented):
I would like to put this in today. The yaml changes look good as far as I can tell. One question: @ackerlar have you tried if the model still functions normally if you use this default setting with a version that does not have iceberg code inside? Otherwise it will be (I fear) a very large struggle to get everyone onto the correct version on the Fortran side. That's already hard enough on the esm side, but if we need to both sides at once it will be chaotic

ackerlar commented:
@pgierz No, I haven't. I guess I need to switch to Ollie for this as I am in the queue on mitral for two days. I will do this today.

ackerlar commented:
I wanted to test this in a virtual environment and installed the esm_tools within this direnv. But it seems that all the config files still come from my home directory when submitting a job...

mandresm commented:
> I would like to put this in today. The yaml changes look good as far as I can tell. One question: @ackerlar have you tried if the model still functions normally if you use this default setting with a version that does not have iceberg code inside? Otherwise it will be (I fear) a very large struggle to get everyone onto the correct version on the Fortran side. That's already hard enough on the esm side, but if we need to both sides at once it will be chaotic

I am certain that this will break at least AWICM3, so please don't merge it yet. The things I mentioned above need to be taken care of before merging and I want to run the testing (actually do some runs) of the whole `esm_tests` before merging, at least in ollie.

**TODO**
- [x] Remove all the icebergs configurations from `fesom-2.0.yaml` (icebergs is only supported in 2.1 versions) @ackerlar I have done that in my local copy, can I push it?
- [x] Put the namelist changes under a version choose (as far as I understand FESOM2.1 main branch does not include icebergs therefore, those changes will make FESOM standalone break)
- [x] Test everything

mandresm committed:
removing icebergs from fesom-2.0.yaml

mandresm commented:
> @mandresm To my understanding, they shall go into the main branch. They are already merge into [[https://gitlab.dkrz.de/FESOM/fesom2/-/commits/paleodyn/awiesm-2.1-gamma](https://github.com/esm-tools/esm_tools/pull/url)], which is, again from my understanding, the FESOM version we are aiming to merge everything into. Icebergs, wiso and pico are already in there, the only thing missing is recom.

I don't see `awiesm-2.1-gamma` version in the yaml files. @ackerlar would you let me know how are you exactly compiling and running fesom with icebergs?

mandresm committed:
suggestions to avoid conflicts with FESOM2.1 versions that have no icebergs

mandresm committed:
fixed bug on a choose block

ackerlar commented:
@mandresm I changed the FESOM version "by hand" and compiled afterwards. Shall I include the gamma version into `fesom-2.1.yaml` and `awiesm-2.1.yaml`?

And sorry,  but the job in the venv did not have been submitted, yet.

mandresm commented:
@ackerlar this is entirely up to you, I don't have any problem with merging this into `release` even if there is not a default AWIESM version with icebergs on it. You are currently developing, and changes will happen in AWIESM to improve the versioning there, so I don't think a missing AWIESM version should be blocking your work. Once #603 is discussed and merged, and once all the `esm_tests` pass I'm happy to merge this into `release`.

ackerlar committed:
Merge pull request #603 from esm-tools/fix/icb_fesom_namelist_suggestions

Fix/icb fesom namelist suggestions

mandresm commented:
Thanks @ackerlar , I'll start a `esm_tests` now in mistral, hopefully by Monday we know if anything breaks. Could you test that my last changes did not break anything for you?

mandresm committed:
Merge remote-tracking branch 'origin/release' into fix/icb_fesom_namelist

ackerlar committed:
add awiesm-2.1-gamma

ackerlar committed:
Merge branch 'fix/icb_fesom_namelist' of https://github.com/esm-tools/esm_tools into fix/icb_fesom_namelist

ackerlar committed:
remove resouces

ackerlar commented:
I added `awiesm-2.1-gamma` and everything works fine for using a FESOM branch without icebergs or one with icebergs. The only thing missing were the iceberg related FESOM output files. I don't really know how to include them in the config file without the `outputs` section.

mandresm commented:
I thought we agreed on discontinuing the `gamma` version of AWI-ESM2.1. Is that incorrect?

About the `esm_tests` `resources`, was that creating a problem for you Lars? I'd would prefer to keep the `resources` submodule around, as it is going to be pretty useful for the ESM-Tools team, but I am very interested in all the problems you might be facing.

mandresm commented:
> I added `awiesm-2.1-gamma` and everything works fine for using a FESOM branch without icebergs or one with icebergs. The only thing missing were the iceberg related FESOM output files. I don't really know how to include them in the config file without the `outputs` section.

About the output files, can you provide the name of those files? Then I'll commit to this branch with that.

mandresm commented on configs/components/fesom/fesom-2.1.yaml:
```diff
@@ -154,14 +167,6 @@ outdata_targets:
-            add_restart_out_files:
-                icb_restart: icb_restart
-                icb_restart_ISM: icb_restart_ISM
-
```
Why are this files deleted here?

ackerlar replied:
I added them at the top under "choose_version" but realized this would be a bad choice and added them back under "choose_use_icebergs"

mandresm committed:
reintroduce esm_tests resources

mandresm commented:
@ackerlar , I have reintroduced the `resources` submodule for `esm_tests`. You can now pull and check if you get any conflicts, hopefully not.

ackerlar committed:
put iceberg restart files behind flag

ackerlar committed:
add iceberg output to default namelist.io

ackerlar commented:
Thanks @mandresm ! I have put the fesom restart files behind the icb flag again. Simulations run for `awiesm-2.1-gamm` and `wiso`. Some issues still remain though:
- in order to get iceberg output, I had to modify `namelist.io` and add `icb` as output field. I guess, this should not be default, right?
- if you don't want to run icebergs with a FESOM version containig the iceberg model, you still have to say `icb_code: true` (nothing else; by default the iceberg model itself is switched off, but the `icebergs` section stilll needs to be added to `namelist.config` as FESOM is looking for it. I would need to change FESOM itself and restructure the iceberg related input to the configs). If you run a FESOM version, that does not contain iceberg code (e.g. wiso), you switch `icb_code: False` and everything runs fine
- the tidying of the experiments takes very long and runs into time out errors (e.g. here: /work/ollie/lackerma/glacial-inception/experiments/test09/log/test09_awiesm_compute_20010101-20011231_14039023.log). Maybe this is related to the very long time it takes for the job to be submitted? For me this only happens when working with direnv but it may take an hour or so just for job submission.

mandresm commented:
>     * in order to get iceberg output, I had to modify `namelist.io` and add `icb` as output field. I guess, this should not be default, right?
> 

I wonder if having the following under a `choose_use_icebergs` would help:
```
    add_namelist_changes:
        namelist.io:
            nml_list:
                io_list(:): ["icb", "1", "d", 4] # Doesn't (:) append to the same name in the namelist once it's read by FESOM?
```

>     * if you don't want to run icebergs with a FESOM version containig the iceberg model, you still have to say `icb_code: true` (nothing else; by default the iceberg model itself is switched off, but the `icebergs` section stilll needs to be added to `namelist.config` as FESOM is looking for it. I would need to change FESOM itself and restructure the iceberg related input to the configs). If you run a FESOM version, that does not contain iceberg code (e.g. wiso), you switch `icb_code: False` and everything runs fine
> 

`icb_code` is a variable that depends on the FESOM version used, and therefore it can be used **automatically** (no user control needed) be including it into the corresponding FESOM version for example:
```
  2.1-gamma:
    branch: paleodyn/awiesm-2.1-gamma
    add_restart_in_files:
        wiso_restart: wiso_restart
    add_restart_out_files:
        wiso_restart: wiso_restart
    add_coupling_fields:
        "[[wiso_fields-->FIELD]]":
            grid: feom
    icb_code: True
```
So I just need to know which versions of FESOM currently use the iceberg code, my guess is that is only `awiesm-2.1-gamma` right @ackerlar ?

>     * the tidying of the experiments takes very long and runs into time out errors (e.g. here: /work/ollie/lackerma/glacial-inception/experiments/test09/log/test09_awiesm_compute_20010101-20011231_14039023.log). Maybe this is related to the very long time it takes for the job to be submitted? For me this only happens when working with direnv but it may take an hour or so just for job submission.

It is very unfortunate, but I currently don't have time to support `direnv` (it's not integrated in ESM-Tools anyway). We are going to take care of it soon, and @pgierz might be able to help you with it. I'm interested in any talk you might have about it, I just don't have time now to dive deep into how to make advance use of it. However, I can offer to run some tests for you (I don't have any simulation going on right now) if you provide a runscript. I will also include that runscript in `esm_tests` so we make sure that it always work before merging anything to release.

ackerlar committed:
icb_code automatically set by fesom version

ackerlar commented:
@mandresm Thanks! Your suggestions worked! The long waiting times don't bother me at the moment as I am using direnv only for testing. 
One thing is, that with your suggested changes, esm_tools don't find the fesom namelist,, as they look in a directory `esm_tools/namelists/fesom2/2.1-gamma` which is not present. So I prescribed the namelist dir in my runscript. This would not be an issue once gamma is the default 2.1 fesom version.

pgierz commented:
To really figure out the `direnv` overhead I need a way to cleanly benchmark the recipe steps. I think that is started (but might not work yet?) in #552

I get the impression this might be happing sporadically for one, and that it has something to do with the fact that `direnv` starts a sub shell whenever you enter a folder. However, in other projects I work on, the overhead caused by it is minimal. Once the profiling works again I can (maybe) say something more definitive......

mandresm reviewed (approved):
Hi @ackerlar , sorry that it took so long to review this. All esm_tests checks passed on my side, in principle we could merge this in `release` right now. However, we could merge this into the bigger `feat/awiesm_final` that we are putting together and make sure these changes are all compatible with the changes there. I can take care of that merge. What do you think?

I'll need a working runscript with icebergs that I could add to `esm_tests`.

mandresm commented:
@ackerlar are you okay with me deleting the gamma version and substituting it by `2.1-awiesm-2.1`? I've seen that `paleodyn/awiesm-2.1-gamma` was merged into `2.1-awiesm-2.1`.

ackerlar committed:
remove default ib_num

ackerlar committed:
update namelists.py

ackerlar commented:
@mandresm Thank you very much! I reopened this because I pushed some changes (hopefully the last ones ...). 
A test runscript can be found here: `/work/ollie/lackerma/glacial-inception/experiments/test_ib06/scripts//awiesm.yaml`

I am totally fine with:
- substituting `paleodyn/awiesm-2.1-gamma`
- merging this branch into `feat/awiesm_final `

mandresm commented:
Thanks so much @ackerlar! I'll be adding the scrip to `esm_tests`, and merging the new changes into `feat/awiesm_final`.

mandresm commented:
Hi again @ackerlar , one thing I'm noticing in your runscript is that you are using different numbers for `nproca` and `nprocb`. I thought there was an advice against this, I'm I wrong @pgierz ?

ackerlar commented:
Thanks @mandresm I changed it.

pgierz commented:
it should always be 24x24 if I remember correctly.

mandresm commented:
Hi @ackerlar ,

I have been trying to reproduce your sample simulation with `feat/awiesm_final` branch and FESOM `2.1-awiesm-2.1` here: 
`/work/ollie/mandresm/testing/run/awiesm/awiesm-2.1_icebergs`

I've compare our namelists and the only one that has major differences is `namelist.echam`. This is because you are using your own `namelist_dir` hardcoded somewhere right?

ackerlar commented:
@mandresm No, I don't use my own namelist for echam. This is the source for namelist.echam
`/work/ollie/lackerma/glacial-inception/software/github.com/esm-tools/esm_tools/namelists/.//echam/6.3.05p2-concurrent_radiation-paleodyn/PI-CTRL/namelist.echam`

pgierz commented:
Has this been superseded by #623?

pgierz commented:
Has this been superseded by #623?

mandresm commented:
> Has this been superseded by #623?

Correct. We can close it now.
